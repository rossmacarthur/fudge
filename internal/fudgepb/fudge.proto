syntax = "proto3";

package fudge;

option go_package = "github.com/rossmacarthur/fudge/internal/fudgepb";

message Error {
    repeated Hop hops = 1;
}

message Hop {
    int32 kind = 1;
    string binary = 2;
    string message = 3;
    string code = 4;
    repeated Frame trace = 5;
}

message Frame {
    string file = 1;
    string function = 2;
    int32 line = 3;
    string message = 4;
    repeated KeyValue key_values = 5;
}

message KeyValue {
    string key = 1;
    string value = 2;
}
