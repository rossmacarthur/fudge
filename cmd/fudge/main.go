package main

import (
	"crypto/rand"
	"fmt"
	"go/token"
	"math"
	"math/big"
	"os"
	"path/filepath"

	"github.com/dave/dst"
	"github.com/dave/dst/decorator"
	"github.com/rossmacarthur/fudge"
	"github.com/rossmacarthur/fudge/errors"
)

const imp = "github.com/rossmacarthur/fudge/errors"

func main() {
	fset := token.NewFileSet()

	err := filepath.Walk(".", func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		if info.IsDir() {
			return nil
		}
		if filepath.Ext(path) != ".go" {
			return nil
		}
		err = rewriteSentinelErrors(fset, path)
		if err != nil {
			return errors.Wrap(err, "", fudge.KV("path", path))
		}
		return nil
	})
	if err != nil {
		fmt.Fprintf(os.Stderr, "error: %#v\n", err)
		os.Exit(1)
	}
}

func rewriteSentinelErrors(fset *token.FileSet, path string) error {
	// Parse the file
	f, err := decorator.ParseFile(fset, path, nil, 0)
	if err != nil {
		return errors.Wrap(err, "")
	}

	// Find and rewrite the sentinel errors
	rewriteSentinelErrorsInFile(f)

	// Write the file back
	of, err := os.Create(path)
	if err != nil {
		return errors.Wrap(err, "")
	}
	defer of.Close()

	err = decorator.Fprint(of, f)
	if err != nil {
		return errors.Wrap(err, "")
	}

	return err
}

func rewriteSentinelErrorsInFile(f *dst.File) {
	var alias string

	dst.Inspect(f, func(n dst.Node) bool {
		if spec, ok := n.(*dst.ImportSpec); ok {
			if spec.Path.Value == `"`+imp+`"` {
				if spec.Name != nil {
					alias = spec.Name.Name
				} else {
					alias = filepath.Base(imp)
				}
			}
		}

		if spec, ok := n.(*dst.ValueSpec); ok {
			// Check if the value is a call to errors.Sentinel
			if len(spec.Values) != 1 {
				return false
			}
			call, ok := spec.Values[0].(*dst.CallExpr)
			if !ok {
				return false
			}
			sel, ok := call.Fun.(*dst.SelectorExpr)
			if !ok {
				return false
			}
			if sel.Sel.Name != "Sentinel" {
				return false
			}
			if sel.X.(*dst.Ident).Name != alias {
				return false
			}

			if len(call.Args) == 2 {
				if lit, ok := call.Args[1].(*dst.BasicLit); ok && lit.Kind == token.STRING {
					if checkCode(lit.Value[1 : len(lit.Value)-1]) {
						return false
					}
				}
			}

			var args []dst.Expr
			if len(call.Args) > 0 {
				args = append(args, call.Args[0])
			} else {
				args = append(args, &dst.BasicLit{
					Kind:  token.STRING,
					Value: `""`,
				})
			}
			call.Args = append(args, &dst.BasicLit{
				Kind:  token.STRING,
				Value: `"` + randomCode() + `"`,
			})

		}

		return true
	})
}

var randReader = rand.Reader

func randomCode() string {
	n, err := rand.Int(randReader, big.NewInt(int64(math.MaxInt64)))
	if err != nil {
		panic(err)
	}
	return fmt.Sprintf("ERR_%016x", n.Int64())
}

func checkCode(code string) bool {
	if len(code) != 20 {
		return false
	}
	if code[:4] != "ERR_" {
		return false
	}
	for _, c := range code[4:] {
		if c < '0' || c > 'f' {
			return false
		}
	}
	return true
}
