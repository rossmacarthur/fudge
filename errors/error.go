package errors

import (
	"fmt"
	"strings"
)

// Error is a concrete error type containing a stack trace
type Error struct {
	// Binary is the name of the executable the error occurred in
	Binary string
	// Message is the optional sentinel message (can be empty)
	Message string
	// Code is the optional sentinel error code (can be empty)
	Code string
	// Cause is the original non-Fudge error (can be nil)
	Cause error
	// Trace is the stack trace
	//
	// contextual messages and key values are attached to individual stack frames
	Trace []Frame
}

// Unwrap implements the errors.Unwrap interface and returns the original
// error if possible
func (e *Error) Unwrap() error {
	return e.Cause
}

// clone deep copies the error
func (e *Error) clone() *Error {
	c := *e
	c.Trace = make([]Frame, 0, len(e.Trace))
	for _, f := range e.Trace {
		c.Trace = append(c.Trace, *f.clone())
	}
	return &c
}

// Is implements the errors.Is interface
//
// A Fudge error is the same as another if they have the same error code. This
// means that you can only compare to a sentinel error.
func (e *Error) Is(target error) bool {
	t, ok := target.(*Error)
	if !ok {
		return false // errors.Is will Unwrap and compare to original
	}

	if e.Code != "" {
		return e.Code == t.Code
	}

	return e == t
}

func (e *Error) Error() string {
	return fmt.Sprintf("%s", e)
}

// Format implements the fmt.Formatter interface
//
// The following verbs are supported:
//
//	%v, %s: print the wrapping messages and error message
//	%+v, %+s: print the error message and stack trace with wrapping messages
//	%#v, %#s: print the error message and stack trace with wrapping messages and key values
func (e *Error) Format(s fmt.State, verb rune) {
	switch verb {
	case 'v', 's':
		fmt.Fprintf(s, "%s", e.fullMessage())

		switch {
		case s.Flag(int('+')):
			for _, f := range e.Trace {
				fmt.Fprint(s, "\n")
				f.Format(s, verb)
			}
		case s.Flag(int('#')):
			kvs := e.fullKeyValues()
			if len(kvs) > 0 {
				fmt.Fprintf(s, " {%v}", kvs)
			}
			for _, f := range e.Trace {
				fmt.Fprint(s, "\n")
				f.Format(s, verb)
			}
		}
	default:
		fmt.Fprintf(s, "%%!%c(*errors.Error=%s)", verb, e.fullMessage())
	}
}

// FormatCustom visits each frame context message in reverse order, then the
// primary error message and then finally the cause.
//
// This is typically what you want to do when formatting the error in a custom
// way. For more complex formatting you can manually implement this function as
// all fields on Error are public.
func (e *Error) FormatCustom(visit func(message string)) {
	for i := len(e.Trace) - 1; i >= 0; i-- {
		m := e.Trace[i].Message
		if m != "" {
			visit(m)
		}
	}
	if e.Message != "" {
		visit(e.Message)
	}
	if e.Cause != nil {
		visit(e.Cause.Error())
	}
}

// fullMessage returns the full error message
func (e *Error) fullMessage() string {
	var s strings.Builder

	write := func(m string) {
		if s.Len() > 0 {
			s.WriteString(": ")
		}
		s.WriteString(m)
	}

	for i := len(e.Trace) - 1; i >= 0; i-- {
		m := e.Trace[i].Message
		if m != "" {
			write(m)
		}
	}

	if e.Message != "" {
		write(e.Message)
		if e.Code != "" {
			s.WriteString(" (")
			s.WriteString(e.Code)
			s.WriteString(")")
		}
	}
	if e.Cause != nil {
		write(e.Cause.Error())
	}

	return s.String()
}

// fullKeyValues returns the full key values
func (e *Error) fullKeyValues() KeyValues {
	kvs := make(KeyValues)
	for i := len(e.Trace) - 1; i >= 0; i-- {
		for k, v := range e.Trace[i].KeyValues {
			kvs[k] = v
		}
	}
	return kvs
}
