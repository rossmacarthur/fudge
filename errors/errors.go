package errors

import (
	"errors"

	"github.com/rossmacarthur/fudge"
)

// New creates a new error with a message and options.
//
// This method is intended to be used to define local errors. A stack trace is
// is attached to the error.
func New(msg string, opts ...fudge.Option) error {
	errors := Error{trace: trace(1)}
	frame := &errors.trace[0]
	frame.message = msg
	for _, o := range opts {
		o.Apply(frame)
	}
	return &errors
}

// NewSentinel creates a new sentinel error with a message and code.
//
// This method is intended to be used to define global sentinel errors. No
// stack trace is attached until these errors are wrapped.
func NewSentinel(msg string, code string) error {
	return &Error{message: msg, code: code}
}

// Wrap wraps an existing error with a new message and options.
//
// This method consumes the original error, so it should not be used afterward.
func Wrap(err error, msg string, opts ...fudge.Option) error {
	if err == nil {
		return nil
	}

	errors, ok := err.(*Error)
	if ok && errors.trace == nil {
		// wrapping a sentinel Fudge error
		errors = errors.Clone()
		errors.trace = trace(1)
		frame := &errors.trace[0]
		frame.message = msg
		for _, o := range opts {
			o.Apply(frame)
		}

	} else if ok {
		// wrapping a Fudge error
		errors = errors.Clone()
		frame := findCallSite(errors, 1)
		frame.message = msg
		for _, o := range opts {
			o.Apply(frame)
		}

	} else {
		// wrapping a non-Fudge error
		errors = &Error{original: err, trace: trace(1)}
		frame := &errors.trace[0]
		frame.message = msg
		for _, o := range opts {
			o.Apply(frame)
		}
	}

	return errors
}

func findCallSite(e *Error, skip int) *Frame {
	file, line := call(skip + 1)
	for i, f := range e.trace {
		if f.file == file && f.line == line {
			return &e.trace[i]
		}
	}

	// the call site doesn't exist in the trace so we need to add it
	// by combining the traces
	trace := trace(skip + 1)
outer:
	for _, f := range trace {
		for j, g := range e.trace {
			if f.file == g.file && f.line == g.line {
				e.trace = append(e.trace[:j], trace...)
				break outer
			}
		}
	}

	for i, f := range e.trace {
		if f.file == file && f.line == line {
			return &e.trace[i]
		}
	}

	panic("failed to find call site frame")
}

// Unwrap is the same as the standard library's errors.Unwrap.
func Unwrap(err error) error {
	return errors.Unwrap(err)
}

// Is is the same as the standard library's errors.Is.
func Is(err error, target error) bool {
	return errors.Is(err, target)
}

// As is the same as the standard library's errors.As.
func As(err error, target any) bool {
	return errors.As(err, target)
}
