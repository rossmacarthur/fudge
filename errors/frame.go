package errors

import (
	"fmt"
	"strings"

	"github.com/go-stack/stack"
)

// Frame is a single frame in a stack trace
type Frame struct {
	// original is the original non-Fudge error (can be nil)
	original error
	// message is the message associated with the frame (can be empty)
	message string
	// keyValues is a map of key-value pairs associated with the frame (can be nil)
	keyValues KeyValues
	// file is the file name of the frame
	file string
	// line is the line number of the frame
	line int
}

// SetKeyValue implements the fudge.apply interface
func (f *Frame) SetKeyValue(k, v string) {
	if f.keyValues == nil {
		f.keyValues = make(map[string]string)
	}
	f.keyValues[k] = v
}

func (f *Frame) String() string {
	return fmt.Sprintf("%v", f)
}

func (f Frame) Format(s fmt.State, verb rune) {
	fmt.Fprintf(s, "%s:%d", f.file, f.line)
	if f.message != "" {
		fmt.Fprintf(s, ": %s", f.message)
	}
	if f.original != nil {
		fmt.Fprintf(s, ": %s", f.original.Error())
	}
	if len(f.keyValues) > 0 && (s.Flag(int('#')) || s.Flag(int('+'))) {
		fmt.Fprintf(s, " {%v}", f.keyValues)
	}
}

func trace(skip int) []Frame {
	var trace []Frame
	for _, c := range stack.Trace()[skip:] {
		f := c.Frame()
		trace = append(trace, Frame{
			file: pkgFilePath(f.Function, f.File),
			line: f.Line,
		})
	}

	return trace
}

func call(skip int) (string, int) {
	c := stack.Caller(skip)
	f := c.Frame()
	return pkgFilePath(f.Function, f.File), f.Line
}

func pkgFilePath(function, file string) string {
	const pathSep = "/"

	var pre, post string

	lastSep := strings.LastIndex(file, pathSep)
	if lastSep == -1 {
		post = file
	} else {
		post = file[strings.LastIndex(file[:lastSep], pathSep)+1:]
	}

	end := strings.LastIndex(function, pathSep)
	if end == -1 {
		return post
	}
	pre = function[:end]

	return pre + pathSep + post
}
