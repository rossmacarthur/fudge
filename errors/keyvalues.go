package errors

import (
	"fmt"
	"sort"
)

type KeyValues map[string]string

func (m *KeyValues) Clone() KeyValues {
	clone := make(KeyValues, len(*m))
	for k, v := range *m {
		clone[k] = v
	}
	return clone
}

func (m *KeyValues) String() string {
	return fmt.Sprintf("%s", m)
}

func (m KeyValues) Format(s fmt.State, verb rune) {
	var keys []string
	for k := range m {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	for i, k := range keys {
		if i > 0 {
			fmt.Fprintf(s, ", %s:%s", k, m[k])
		} else {
			fmt.Fprintf(s, "%s:%s", k, m[k])
		}
	}
}
