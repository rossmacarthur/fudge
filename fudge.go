package fudge

type Option interface {
	Apply(apply)
}

type apply interface {
	SetKeyValue(k, v string)
}

type kv struct {
	k, v string
}

func (o *kv) Apply(a apply) {
	a.SetKeyValue(o.k, o.v)
}

func KV(k, v string) Option {
	return &kv{k, v}
}
