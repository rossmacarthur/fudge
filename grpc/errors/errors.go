package grpcerrors

import (
	"context"

	"google.golang.org/grpc"
)

// UnaryClientInterceptor is a gRPC client interceptor that converts gRPC status
// details into Fudge errors.
func UnaryClientInterceptor(ctx context.Context, method string, req, resp any,
	cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {

	err := invoker(ctx, method, req, resp, cc, opts...)
	return interceptClient(err)
}

// UnaryServerInterceptor is a gRPC server interceptor that returns Fudge error
// information in the gRPC status details.
func UnaryServerInterceptor(ctx context.Context, req any,
	info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (any, error) {

	resp, err := handler(ctx, req)
	return resp, interceptServer(err)
}
