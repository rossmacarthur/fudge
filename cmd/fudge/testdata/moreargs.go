package test

import "github.com/rossmacarthur/fudge/errors"

var ErrTest = errors.Sentinel("test error", "ERR_12345", "test")
