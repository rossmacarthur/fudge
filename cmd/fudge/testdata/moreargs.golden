package test

import "github.com/rossmacarthur/fudge/errors"

var ErrTest = errors.Sentinel("test error", "ERR_1566c74d10037c4d")
